### PR TITLE
[query] Nuke streamable

### PR DESCRIPTION
**Pull request by cseed**
Summary of changes:
 - removed TStreamable
 - All IR inputs must be TStream or TArray, statically
 - renamed Array* nodes to be Stream* (e.g. ArrayMap to StreamMap, etc.)
 - ToArray, ToDict and ToSet take TStream only
 - LowerArrayToStream is gone.  This effectively happens in the front end.

Most of the code is peppering ToArray and ToStream in the right places.

I think this is a nice improvement in two ways:

I makes the IR more transparent by making ToArray, a potentially expensive operation, completely explicit.  (Uses of ToArray should probably be audited.)

Second, I think it cleans up the flow in Emit/Stream.  Now, streams are always compiled by EmitStream, and non-stream values that correspond to EmitTriplets in the code are compiled by Emit only.

Finally, I fixed a bug in array sorter that would throw an assert on sets with NaNs (compare false to themselves).  I'm a little surprised this isn't failing any tests in master.

### CODE REVIEW

**Comment by cseed** [comment]
I was confused about something which your comment clarified, thanks.  Yeah, my change is incredibly bad.  I need a story about free conversion from dict/set to array.  ToArray playing the dual role is confusing, because sometimes it is very expensive and sometimes nearly free, which isn't a great.

I think I will add another node: ContainerToArray.

File: hail/python/hail/expr/expressions/base_expression.py [hunk 1]
```diff
@@ -281,7 +281,7 @@ def _to_expr(e, dtype):
                      else hl.literal(element, dtype.element_type)
                      for element in elements]
             indices, aggregations = unify_all(*exprs)
-            ir = ToSet(MakeArray([e._ir for e in exprs], None))
+            ir = ToSet(ToStream(MakeArray([e._ir for e in exprs], None)))
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/python/hail/expr/expressions/base_expression.py]: Yes, that's a good suggestion.  There are two issues: the new stream emit stuff doesn't handle wrapping, so large MakeStreams are going to fail for the time being.  Second, the code isn't quite as good, MakeStream creates a auxiliary variable to sore which element its on and does a switch to figure out which element to produce next.  On the other hand, MakeArray directly constructs the array in place.

So for now I think we should prefer MakeArray to MakeStream.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 2]
```diff
@@ -602,13 +601,7 @@ private class Emit(
...
 
       case ToArray(a) =>
-        a.pType match {
-          case typ: PStream =>
-            EmitStream2.toArray(mb, typ.asPArray, emitStream2(a))
-          case _ => emit(a)
-        }
-
-      case ToStream(a) => emit(a)
+        emitArrayIterator(a).toEmitTriplet(mb, PArray(coerce[PStreamable](ir.pType).elementType))
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I'm planning on doing a second pass to rip things out related to PStreamable.

**Comment by cseed** [comment]
> I think we can also get rid of TIterable by refining the type casting in ToSet/ToDict/ToArray/ToStream nodes to reflect the new semantics.

Yes, I will consider that in a later pass.

 > For python---the idea is that we're not really introducing any user-visible changes, right?

Correct.  And yes, the optimizer should strip out any glue.

 > I'm a little nervous about the ToArray(ToStream(dict/set) idiom we're using here

OK, this is fixed.  I added a CastToArray which takes a TContainer and does a no-op cast to array type.  I use it wherever I can and I added a simplify rule to introduce it when possible.

**Comment by cseed** [comment]
Some more comments on this array casting business:
 - We could have made Array{Ref, Len} work on any container.  That would fix most uses of CastToArray, but it would still leave out the cast that we genuinely want to convert a set/dict to an array.  I don't think we ever do that, but a user could with hl.array.
 - It is partially an accident that dict, set and array use the same representation.  We could imagine others, e.g. a hashtable for set.  In which case, the cast wouldn't work, the ArrayLen probably would, and ArrayRef might or might not.
 - Array{Ref, Len} could apply to streams, too (they'll need custom codegen tho).

**Comment by catoverdrive** [comment]
Looks good! 
- There's a couple of places where we use `hl.array()` for our own methods, e.g. `Table.to_matrix_table` and the vcf_combiner that I think are just conceptually "turn this set/dict into an array, please" as opposed to casting out of convenience (like we did for filter, etc. before this PR).
- I think as we add more representations for array/set/dict the CastToArray logic will just need to take into account the pType that it's casting to/from accordingly; it'll therefore still "work" but won't always return a sorted array or guarantee an ordering (which I think we shouldn't consider part of the semantics anyways)
- I don't know that ArrayRef makes sense on general Set/Dict objects directly anyways, since they don't conceptually have a well-defined element order (I think a lot of languages don't let you index into sets/dicts by index, right?). I think I'd agree that ArrayLen should work on everything.
- I'm kind of in favor of introducing companion StreamRef/StreamLen nodes for ArrayRef/ArrayLen to work on streams, just because they'll look very distinct in the generated code.

**Pull request merged**